File: python/src/patterns/iterator/concrete_aggregate.py
from __future__ import annotations

from typing import TYPE_CHECKING

from .concrete_iterator import ConcreteIterator

if TYPE_CHECKING:
    from .iterator import Iterator


class ConcreteAggregate:
    """Holds the items and hands out iterators over them."""

    def __init__(self) -> None:
        self._items: list[str] = []

    def add(self, item: str) -> None:
        self._items.append(item)

    def count(self) -> int:
        return len(self._items)

    def _get(self, index: int) -> str:
        """Access for the iterator; the list itself is never exposed."""
        return self._items[index]

    def create_iterator(self) -> Iterator:
        return ConcreteIterator(self)

File: javascript/src/iterator/concrete-aggregate.js
// Iterator: access an aggregate's elements sequentially without exposing its representation.
// An iterator is any object with hasNext() and next(); an aggregate any object with createIterator().

import { ConcreteIterator } from "./concrete-iterator.js";

/** Holds the items and hands out iterators over them. */
export class ConcreteAggregate {
  #items = [];

  add(item) {
    this.#items.push(item);
  }

  count() {
    return this.#items.length;
  }

  /** Access for the iterator; the array itself is never exposed. */
  get(index) {
    return this.#items[index];
  }

  createIterator() {
    return new ConcreteIterator(this);
  }
}
